### PR TITLE
Check for improperly formatted arguments (fix #885)

### LINKED ISSUE #885: Argument parsing gets confused by -arg-name instead of --arg-name
### Description

Name argument gets interleaved into electrum download path

### Steps to reproduce issue
```
$ npm install nativefier -g
$ nativefier -name "Google Voice" "https://voice.google.com/"
Downloading tmp-3522-0-electron-vGoogle Voice-darwin-x64.zip
Error: GET https://github.com/electron/electron/releases/download/vGoogle Voice/electron-vGoogle Voice-darwin-x64.zip returned 404
Error: Failed to find Electron vGoogle Voice for darwin-x64 at https://github.com/electron/electron/releases/download/vGoogle Voice/electron-vGoogle Voice-darwin-x64.zip
```

### Workaround
Don't use the name argument.

### PR DESCRIPTION
**Pull request by sorhtyre**
Verifies short arguments do not contain an extra dash ( right: `-h`, error: `--h` )
Verifies long arguments contain at least two dashes ( right: `--help`, error: `-help` )

### CODE REVIEW

**Review by ronjouch — CHANGES_REQUESTED** [review]
Thanks @sorhtyre ! A few things:

File: src/cli.ts [hunk 1]
```diff
@@ -68,6 +68,14 @@ function checkInternet(): void {
 if (require.main === module) {
   const sanitizedArgs = [];
   process.argv.forEach((arg) => {
+    if (
+      !arg.search(/^-[A-Z]{2,}/i) ||
+      (!arg.search(/^--[A-Z]/i) && arg.length == 3)
+    ) {
+      log.error(`Invalid argument passed: ${arg} \nAborting`);
```
[the remark this answers is not in the document]
Comment (sorhtyre) [reply src/cli.ts]: :+1:

File: src/cli.ts [hunk 2]
```diff
@@ -68,6 +68,14 @@ function checkInternet(): void {
 if (require.main === module) {
   const sanitizedArgs = [];
   process.argv.forEach((arg) => {
+    if (
+      !arg.search(/^-[A-Z]{2,}/i) ||
+      (!arg.search(/^--[A-Z]/i) && arg.length == 3)
```
[the remark this answers is not in the document]
Comment (sorhtyre) [reply src/cli.ts]: :+1:

File: src/cli.ts [hunk 3]
```diff
@@ -68,6 +68,14 @@ function checkInternet(): void {
...
+    ) {
+      log.error(`Invalid argument passed: ${arg} \nAborting`);
+      process.exit(1);
+    }
```
[the remark this answers is not in the document]
Comment (sorhtyre) [reply src/cli.ts]: I am not. Let me do some research.

[the remark this answers is not in the document]
Comment (sorhtyre) [reply src/cli.ts]: If I understand this correctly, you want me to move this into helpers/helpers.ts as an exportable function and create a cli-tests.ts to verify it?

[the remark this answers is not in the document]
Comment (ronjouch) [reply src/cli.ts]: > If I understand this correctly, you want me to move this into helpers/helpers.ts as an exportable function and create a cli-tests.ts to verify it?

@sorhtyre yes! Nit: I see we already have a `helpers.test.ts`, no need to create a cli-tests.

**Comment by sorhtyre** [comment]
Updated as a helper function with an associated unit test.

**Comment by sorhtyre** [comment]
Ok, so trying to import from `app/src/helpers/helpers.ts` creates a mangled filesystem. Should it move the helper function into `src/helpers/helpers.ts` and create a helpers-test.ts there to run the unit test?

File: app/src/helpers/helpers.ts [hunk 4]
```diff
@@ -87,3 +87,8 @@ export function getCounterValue(title: string): string {
   const match = itemCountRegex.exec(title);
   return match ? match[1] : undefined;
 }
+
+export function isArgFormatValid(arg: string): boolean {
```
Comment (ronjouch) [inline app/src/helpers/helpers.ts]: Good naming! 👍

File: app/src/helpers/helpers.test.ts [hunk 5]
```diff
@@ -42,3 +42,29 @@ test('getCounterValue should return a string for small counter numbers in the ti
...
+  expect(isArgFormatValid(extraLongArg)).toEqual(true);
+});
+
+test('isArgFormatValid should return false for improperly formatted short arguments', () => {
+  expect(isArgFormatValid(invalidShortArg)).toEqual(false);
+});
+
+test('isArgFormatValid should return false for improperly formatted long arguments', () => {
+  expect(isArgFormatValid(invalidLongArg)).toEqual(false);
+});
```
Comment (ronjouch) [inline app/src/helpers/helpers.test.ts]: Not your fault, but the initial test suite code lacked sub-suite structure. Test suites are typically structured into sub-suites, one per tested function, which provides a range of benefits:

1. Clear visual/mental separation between sub-suites
1. Proper scoping & isolation of sub-suite constants
2. Easiness to run only one sub-suite (with a little `.only` at sub-suite level)

`helpers.test.ts` tests three function: `linkIsInternal`, `getCounterValue`, and now `isArgFormatValid`. Can you separate group of tests for each into its own sub-suite by wrapping groups of tests into a `describe` call? See `inferUserAgent.test.ts` for an example.

File: src/cli.ts [hunk 6]
```diff
@@ -6,6 +6,7 @@ import * as dns from 'dns';
 import * as log from 'loglevel';
 
 import { buildNativefierApp } from './main';
+import { isArgFormatValid } from '../app/src/helpers/helpers';
```
Comment (ronjouch) [inline src/cli.ts]: Careful with imports automatically-inserted by your editor, they sometimes do weird stuff.

```suggestion
import { isArgFormatValid } from './helpers/helpers';
```

File: app/src/helpers/helpers.ts [hunk 7]
```diff
@@ -87,3 +87,8 @@ export function getCounterValue(title: string): string {
...
+  // eslint-disable-next-line
```
Comment (ronjouch) [inline app/src/helpers/helpers.ts]: I think I saw in a failed CI build that this ignore is because [eslint/prefer-regexp-exec](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/prefer-regexp-exec.md) wants you to use `RegExp.exec`, correct?

If correct, sorry for the successive changes (I already asked you to change from `String.search` to `String.match`), but I'm going to suggest we do yet another change and go with the RegExp.exec recommended by ESLint. Ignoring the recommendation wouldn't be dramatic since performance is moot here (we're talking about a few meager calls), but switching to the alternative is fine-looking too, so I'd rather have it than an easily-avoided eslint-ignore.

File: app/src/helpers/helpers.ts [hunk 8]
```diff
@@ -87,3 +87,8 @@ export function getCounterValue(title: string): string {
...
+  return arg.match(/^-[a-z]$/i) || arg.match(/^--[a-z]{2,}(-[a-z]{2,})?$/i) ? true : false;
```
Comment (ronjouch) [inline app/src/helpers/helpers.ts]: Hmmmmmm no, `/^--[a-z]{2,}(-[a-z]{2,})?$/` will break arguments with multiple dashes, like [`--ignore-gpu-blacklist`](https://github.com/jiahaog/nativefier/blob/master/docs/api.md#ignore-gpu-blacklist).

Correct me if I'm wrong (and this is exaaaaaactly the kind of case you should add to your unit tests!), but I think that `/^--[a-z-]{2,}?$/` will work well. It may look weird if you're young with regexes because of the dash between a and z, and the "standalone" dash, but the regex parser will handle it. Try it and verify it with an extra test!

**Review by ronjouch — CHANGES_REQUESTED** [review]
Nice job on the tests!!! Almost there.

File: src/helpers/helpers-test.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import { isArgFormatValid } from './helpers';
+
+const shortArg = '-t';
+const longArg = '--test';
+const extraLongArg = '--test-run';
+const invalidShortArg = '--t';
+const invalidLongArg = '-test';
```
Comment (ronjouch) [inline src/helpers/helpers-test.ts:7]: One more thing: if there's no string reuse, don't bother making `const` of these things, there's no benefit in extracting a const , and

```ts
expect(isArgFormatValid('-t')).toEqual(true);
```

... is actually more readable (since there's one less mental indirection to do) than

```ts
const shortArg = '-t';
expect(isArgFormatValid(shortArg)).toEqual(true);
```

File: app/src/helpers/helpers.ts (hunk 8 above)
Comment (sorhtyre) [reply app/src/helpers/helpers.ts]: I only tested for one additional dash... oops, I'll modify that. 
`[a-z-]` does not work because it incorrectly accepts `---name`

Comment (ronjouch) [reply app/src/helpers/helpers.ts]: > `[a-z-]` does not work because it incorrectly accepts `---name`

@sorhtyre good catch. `/^--[a-z][a-z-]{1,}?$/i` maybe, then?

File: src/helpers/helpers-test.ts (hunk 9 above)
Comment (sorhtyre) [reply src/helpers/helpers-test.ts:7]: I thought about doing that but followed the example in the current `app/src/helpers/helpers.test.ts`
I'll make the change!

File: app/src/helpers/helpers.ts (hunk 7 above)
Comment (sorhtyre) [reply app/src/helpers/helpers.ts]: I'll make the change!

File: src/helpers/helpers-test.ts (hunk 9 above)
Comment (ronjouch) [reply src/helpers/helpers-test.ts:7]: > I thought about doing that but followed the example in the current `app/src/helpers/helpers.test.ts`. I'll make the change!

👍! Yeah, but see: in the `linkIsInternal`, there is actual cross-test reuse of for example the `internalUrl` constant. So in this case, it makes total sense to extract it into a variable, because you don't want to duplicate it.

File: app/src/helpers/helpers.ts (hunk 8 above)
Comment (sorhtyre) [reply app/src/helpers/helpers.ts]: Got it: `/^--[a-z]{2,}(-[a-z]{2,})*$/i` will catch 0 or more single dashes inside a long argument.
As for the unit test, what other cases should I test against?

Comment (ronjouch) [reply app/src/helpers/helpers.ts]: > As for the unit test, what other cases should I test against?

Your current cases are good, I'd just add a test that `--ignore-gpu-blacklist` is considered valid. That's what unit tests are good for: as you iterate on a function, you realize corner case, adjust the implementation, and you enrich your test cases. That way, as the function gets more complex, you ensure you're not regressing by breaking stuff on the left as you're fixing stuff on the right.

File: src/helpers/helpers.test.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
+import { isArgFormatValid } from './helpers';
```
Comment (ronjouch) [inline src/helpers/helpers.test.ts:1]: Ooooooooooooooh, my previous comment must have confused you. I was actually looking at ***`app`*** `/src/helpers/helpers.test.ts` 😄. Alright, well, your code is even better (as more future-proof) with a `describe`.

**Review by ronjouch — APPROVED** [review]
:clap: :clap: :clap: , thanks again @sorhtyre. I pushed a little commit of final nits:

1. In tests, there are many ways to test for equality, with various corner cases. Among the test assertions provided by our test framework Jest, it's best to use the strictest. So, `toEqual` → `toBe`
2. In regex, `[charactergroup]{1}` is redundant, it's equivalent to `[charactergroup]`. And `[singlecharacter]` is redundant, it's equivalent to `[singlecharacter]`. So, `[-]{1}` is better expressed as `-` 🙂.


With that, merging. Will cut a release soon.
